nixos/modules/services/printing/cups-pdf.nix

```
@@ -0,0 +1,205 @@
+
+  # wrapper script that redirects calls to the suid wrapper
+  wrapper = pkgs.writeTextFile {
+    name = "${pkgs.cups-pdf-to-pdf.name}-wrapper.sh";
+    executable = true;
+    destination = "/lib/cups/backend/cups-pdf";
+    checkPhase = ''
+      ${pkgs.stdenv.shellDryRun} "$target"
+      ${getExe pkgs.shellcheck} "$target"
+    '';
+    text = ''
+      #! ${pkgs.runtimeShell}
+      exec "${config.security.wrapperDir}/cups-pdf" "$@"
+    '';
+  };
+
+  # wrapped cups-pdf package that uses the suid wrapper
+  wrapped = pkgs.buildEnv {
+    name = "${pkgs.cups-pdf-to-pdf.name}-wrapped";
+    # using wrapper as first path ensures the wrapper is used
+    paths = [ wrapper pkgs.cups-pdf-to-pdf ];
+    ignoreCollisions = true;
+  };
+
+  instanceSettings = name: {
+    freeformType = nullOr (oneOf [ int str path package ]);
+    # override defaults:
+    # inject instance name into paths,
+    # also avoid conflicts between user names and special dirs
+    options.Out = mkOption {
+      type = nullOr singleLineStr;
+      default = "/var/spool/cups-pdf-${name}/users/\${USER}";
+      defaultText = "/var/spool/cups-pdf-{instance-name}/users/\${USER}";
+      example = "\${HOME}/cups-pdf";
+      description = "output directory";
+    };
+    options.AnonDirName = mkOption {
+      type = nullOr singleLineStr;
+      default = "/var/spool/cups-pdf-${name}/anonymous";
+      defaultText = "/var/spool/cups-pdf-{instance-name}/anonymous";
+      example = "/tmp";
+      description = "path for anonymously created PDF files";
+    };
+    options.Spool = mkOption {
+      type = nullOr singleLineStr;
+      default = "/var/spool/cups-pdf-${name}/spool";
+      defaultText = "/var/spool/cups-pdf-{instance-name}/anonymous";
+      example = "/tmp/cups-pdf";
+      description = "spool directory";
+    };
+    options.Anonuser = mkOption {
+      type = nullOr singleLineStr;
+      default = "";
+      example = "root";
+      description = ''
+        User for anonymous PDF creation.
+        An empty value disables this feature.
+        Avoid the default value <literal>nobody</literal>
+        as it might cause security issues.
+      '';
+    };
+    options.GhostScript = mkOption {
+      type = nullOr path;
+      default = getExe pkgs.ghostscript;
+      defaultText = literalExpression "lib.getExe pkgs.ghostscript";
+      example = literalExpression ''''${pkgs.ghostscript}/bin/ps2pdf'';
+      description = "location of GhostScript binary";
+    };
+  };
+
+  instanceConfig = { name, config, ... }: {
+    options = {
+      enable = (mkEnableOption "this cups-pdf instance")
+               // { default = true; };
+      installPrinter = (mkEnableOption ''
+        a CUPS printer queue for this instance.
+        The queue will be named after the instance and will use
+        the <filename>CUPS-PDF_opt.ppd</filename> ppd file.
+        If this is disabled, you need to add
+        the queue yourself to use the instance
```

```suggestion
        The queue will be named after the instance and will use the <filename>CUPS-PDF_opt.ppd</filename> ppd file.
        If this is disabled, you need to add the queue yourself to use the instance
```
Such short lines are harder to read. We don't need to limit us to 60 characters. There is plenty of space left.